### PR TITLE
MB-62230 - Pre-filtering support for kNN 

### PR DESCRIPTION
**Pull request by metonymic-smokey**
Adds a method, SearchWithIDs() which limits the search to only the IDs part of the inclusion list.

### CODE REVIEW

File: index.go [hunk 1]
```diff
@@ -157,31 +158,25 @@ func (idx *faissIndex) Search(x []float32, k int64) (
...
-	var selector *C.FaissIDSelector
-	if len(exclude) > 0 {
-		excludeSelector, err := NewIDSelectorNot(exclude)
-		if err != nil {
-			return nil, nil, err
-		}
-		selector = excludeSelector.sel
-		defer excludeSelector.Delete()
+	includeSelector, err := NewIDSelectorBatch(include)
+	defer includeSelector.Delete()
```
[the remark this answers is not in the document]
Comment (metonymic-smokey) [reply index.go:164]: right, thanks, fixing this.

File: index.go [hunk 2]
```diff
@@ -157,31 +158,25 @@ func (idx *faissIndex) Search(x []float32, k int64) (
...
 
-	var selector *C.FaissIDSelector
-	if len(exclude) > 0 {
-		excludeSelector, err := NewIDSelectorNot(exclude)
-		if err != nil {
-			return nil, nil, err
-		}
-		selector = excludeSelector.sel
-		defer excludeSelector.Delete()
+	includeSelector, err := NewIDSelectorBatch(include)
```
[the remark this answers is not in the document]
Comment (metonymic-smokey) [reply index.go:163]: include will never be 0 - this is verified at the zapx layer.

File: index.go [hunk 3]
```diff
@@ -157,31 +158,25 @@ func (idx *faissIndex) Search(x []float32, k int64) (
 	return
 }
 
-func (idx *faissIndex) SearchWithoutIDs(x []float32, k int64, exclude []int64, params json.RawMessage) (
-	distances []float32, labels []int64, err error,
+func (idx *faissIndex) SearchWithIDs(x []float32, k int64, include []int64,
```
[the remark this answers is not in the document]
Comment (metonymic-smokey) [reply index.go:160]: Un-filtered queries, for now, will have an include list consisting of all the live docs in the index.
This is still something i'm thinking on how to handle this - for now, i just pass all the live docs as part of the index. If perf tests show it to be a bottleneck, will amend the approach.

File: index.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (moshaad7) [reply index.go:163]: It should be 0 for unfiltered queries

File: index.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (moshaad7) [reply index.go:160]: It doesn't sound appropriate to keep non zero include list for unfiltered queries

File: index.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (metonymic-smokey) [reply index.go:194]: yep, I handle this at the zapx level by checking if a kNN query requires filtering. If not, it's an "unfiltered" query and does the normal search.
Thanks for raising it!

File: index.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (metonymic-smokey) [reply index.go:191]: same response as the other review comment.
I now retain the exclusion list, used in search without IDs for unfiltered queries.

[the remark this answers is not in the document]
Comment (abhinavdangeti) [reply index.go:191]: >Un-filtered queries, for now, will have an include list consisting of all the live docs in the index.

I'm a bit confused but if you're saying that this is now the API that will be used for vector search queries with or without filters - this is not a good idea at all. The list of vectorsToInclude will be typically massive as opposed to vectorsToExclude. So, I think I agree with @moshaad7 here (?).

This method should _only_ be applicable for queries with filters.

**Review by abhinavdangeti — CHANGES_REQUESTED** [review]

**Comment by abhinavdangeti** [comment]
@metonymic-smokey please keep in mind that with the introduction of this capability, we want absolutely no difference in behavior for vector search queries without filters - in terms of compute/allocs or anything else.

File: index.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (metonymic-smokey) [reply index.go:191]: @abhinavdangeti yes, I agree too which is why I only use this when filtering is required(as seen in my zapx PR).

**Review by abhinavdangeti — DISMISSED** [review]

File: index.go [hunk 4]
```diff
@@ -185,6 +188,25 @@ func (idx *faissIndex) SearchWithoutIDs(x []float32, k int64, exclude []int64, p
 	return
 }
 
+func (idx *faissIndex) SearchWithIDs(x []float32, k int64, include []int64,
+	params json.RawMessage) (distances []float32, labels []int64, err error,
+) {
+	includeSelector, err := NewIDSelectorBatch(include)
+	defer includeSelector.Delete()
```
Comment (Thejas-bhat) [inline index.go:195]: maybe this bit got missed, but NewIDSelectorBatch() does return nil when err != nil so I think this should be moved to line > 198

Comment (metonymic-smokey) [reply index.go]: done, thanks!

**Review by abhinavdangeti — APPROVED** [review]

**Pull request merged**